### PR TITLE
Claymanager skeleton

### PR DESCRIPTION
**Pull request by Zemurin**
Claymanager is the geographical manager for the various superregions/regions/states/provinces.
It can now load vanilla state geography (what provinces go into which states), assign terrain to those provinces (relevant for impassables, lakes, oceans - maybe), as well as group states into regions and superregions according to strategic region definitions (/common/strategic_regions/).

This is a skeleton only, for now.

### CODE REVIEW

File: EU4toV3/Source/V3World/ClayManager/TerrainLoader/TerrainLoader.cpp [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+}
+
+void V3::TerrainLoader::registerKeys()
+{
+	registerRegex(commonItems::catchallRegex, [this](const std::string& provinceName, std::istream& theStream) {
+		const auto theTerrain = commonItems::getString(theStream);
+		auto theProvinceName = commonItems::remQuotes(provinceName);
+		std::transform(theProvinceName.begin(), theProvinceName.end(), theProvinceName.begin(), ::toupper);
+		if (theProvinceName.starts_with("X") && theProvinceName.size() == 7)
+			theProvinceName = "x" + theProvinceName.substr(1, theProvinceName.length() - 1);
```
Comment (Idhrendur) [inline EU4toV3/Source/V3World/ClayManager/TerrainLoader/TerrainLoader.cpp:21]: Change some of the test inputs to verify these transformations.

File: EU4toV3/Source/V3World/ClayManager/TerrainLoader/TerrainLoader.cpp [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+		else
+			Log(LogLevel::Warning) << "Encountered terrain for province " << theProvinceName << " in unknown format!";
```
Comment (Idhrendur) [inline EU4toV3/Source/V3World/ClayManager/TerrainLoader/TerrainLoader.cpp:23]: Add test input to verify this warning

File: EU4toV3/Source/V3World/ClayManager/ClayManager.h [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+namespace V3
+{
+class ClayManager
+{
+  public:
+	ClayManager() = default;
+
+	void initializeVanillaStates(const std::string& v3Path);
+	void loadTerrainsIntoProvinces(const std::string& v3Path);
+	void initializeSuperRegions(const std::string& v3Path);
```
Comment (Idhrendur) [inline EU4toV3/Source/V3World/ClayManager/ClayManager.h:17]: Consider using ModFilesystem

File: EU4toV3/Source/V3World/ClayManager/ClayManager.cpp [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+void V3::ClayManager::initializeVanillaStates(const std::string& v3Path)
+{
+	StateLoader stateLoader;
+	stateLoader.loadStates(v3Path);
+	states = stateLoader.getStates();
+
+	auto provinceCount = 0;
+	for (const auto& state: states | std::views::values)
+		provinceCount += static_cast<int>(state->getProvinces().size());
```
Comment (Idhrendur) [inline EU4toV3/Source/V3World/ClayManager/ClayManager.cpp:16]: Use std::accumulate

File: EU4toV3/Source/V3World/ClayManager/ClayManager.cpp [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+				const auto& terrain = terrains.at(provinceName);
+				province->setTerrain(terrain);
+				if (terrain == "ocean")
+					province->setSea();
+				else if (terrain == "lakes")
+					province->setLake();
+			}
+			else
+			{
+				Log(LogLevel::Warning) << "Terrain for province " << provinceName << " cannot be found.";
```
Comment (Idhrendur) [inline EU4toV3/Source/V3World/ClayManager/ClayManager.cpp:38]: Maybe test this

File: EU4toV3/Source/V3World/ClayManager/ClayManager.cpp [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+void V3::ClayManager::initializeSuperRegions(const std::string& v3Path)
+{
+	SuperRegionLoader superRegionLoader;
+	superRegionLoader.loadSuperRegions(v3Path);
+	superRegions = superRegionLoader.getSuperRegions();
+
+	auto regionCount = 0;
+	for (const auto& superRegion: superRegions | std::views::values)
+		regionCount += static_cast<int>(superRegion->getRegions().size());
```
Comment (Idhrendur) [inline EU4toV3/Source/V3World/ClayManager/ClayManager.cpp:50]: Use std:;accumulate

File: EU4toV3/Source/V3World/ClayManager/ClayManager.cpp [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+{
+	for (const auto& superRegion: superRegions | std::views::values)
+		for (const auto& region: superRegion->getRegions() | std::views::values)
+		{
+			std::map<std::string, std::shared_ptr<State>> theStates;
+			for (const auto& stateName: region->getStates() | std::views::keys)
+				if (states.contains(stateName))
+					theStates.emplace(stateName, states.at(stateName));
+				else
+					Log(LogLevel::Warning) << "Attempting to assign state " << stateName << " which doesn't exist to region " << region->getName() << "!";
```
Comment (Idhrendur) [inline EU4toV3/Source/V3World/ClayManager/ClayManager.cpp:64]: Maybe test this

**Comment by Idhrendur** [comment]
Did you purposefully choose V3 over Vic3? It's not a big deal, but with the last game Vic2 was the shorthand that eventually won out.

File: EU4toV3/Source/V3World/ClayManager/StateLoader/State.cpp [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+	registerKeyword("impassable", [this](std::istream& theStream) {
+		const auto theProvinces = commonItems::stringList(theStream).getStrings();
+		for (const auto& provinceName: theProvinces)
+		{
+			auto theProvinceName = commonItems::remQuotes(provinceName);
+			std::transform(theProvinceName.begin(), theProvinceName.end(), theProvinceName.begin(), ::toupper);
+			if (theProvinceName.starts_with("X") && theProvinceName.size() == 7)
+				theProvinceName = "x" + theProvinceName.substr(1, theProvinceName.length() - 1);
+			else
+				Log(LogLevel::Warning) << "Encountered impassable province " << theProvinceName << " in unknown format!";
```
[the remark this answers is not in the document]
Comment (Zemurin) [reply EU4toV3/Source/V3World/ClayManager/StateLoader/State.cpp:40]: Well, it's a two-liner with a variable warning message. I fear a helper function + its surrounding logic would in fact be more verbose and less clear.

Fortunately we load provinces on a precious few places so there won't be many more instances of this check.

**Comment by Zemurin** [comment]
> Did you purposefully choose V3 over Vic3? It's not a big deal, but with the last game Vic2 was the shorthand that eventually won out.

Yes. I'm fine with Vic3 being the name of the converter and how we reference the game itself, but V3 is easier to type and used for the Vic3 portion of the namespace as we did with V2. This isn't an issue for me.

**Pull request merged**